File: src/main/scala/info/crlog/higgs/ssl/SSLContextFactory.scala
package info.crlog.higgs.ssl
package com.fillta.https

import javax.net.ssl.KeyManagerFactory
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManagerFactory
import java.io.FileInputStream
import java.security.KeyStore
import java.security.SecureRandom

object SSLContextFactory {
  def getSSLSocket(sslConfiguration: SSLConfiguration): SSLContext = {
    var useTrustStore: Boolean = false
    var tmf: TrustManagerFactory = null
    var trustStore: KeyStore = null
    try {
      trustStore = KeyStore.getInstance(sslConfiguration.getTrustStoreType)
      if (sslConfiguration.getTrustStorePath != null) {
        trustStore.load(new FileInputStream(sslConfiguration.getTrustStorePath), if (sslConfiguration.getTrustStorePassword == null) "".toCharArray else sslConfiguration.getTrustStorePassword.toCharArray)
        tmf = TrustManagerFactory.getInstance(sslConfiguration.getTrustManagerFactoryType)
        tmf.init(trustStore)
        useTrustStore = true
      }
      else {
        useTrustStore = false
      }
    }
    catch {
      case e: Exception => {
        System.out.println("Unable to create TrustStore Manager. Reason: " + e.getMessage)
      }
    }
    var useClientKeyStore: Boolean = false
    var kmf: KeyManagerFactory = null
    try {
      val clientKeyStore: KeyStore = KeyStore.getInstance(sslConfiguration.getKeyStoreType)
      if (sslConfiguration.getKeyStorePath != null) {
        clientKeyStore.load(new FileInputStream(sslConfiguration.getKeyStorePath), if (sslConfiguration.getKeyStorePassword == null) "".toCharArray else sslConfiguration.getKeyStorePassword.toCharArray)
        kmf = KeyManagerFactory.getInstance(sslConfiguration.getKeyManagerFactoryType)
        kmf.init(clientKeyStore, if (sslConfiguration.getKeyPassword == null) "".toCharArray else sslConfiguration.getKeyPassword.toCharArray)
        useClientKeyStore = true
      }
      else {
        useClientKeyStore = false
      }
    }
    catch {
      case e: Exception => {
        System.out.println("Unable to create KeyStore Manager. Reason: " + e.getMessage)
      }
    }
    try {
      val ctx: SSLContext = SSLContext.getInstance(sslConfiguration.getSecurityProtocol)
      ctx.init(if (useClientKeyStore) kmf.getKeyManagers else null, if (useTrustStore) tmf.getTrustManagers else null, new SecureRandom)
      return ctx
    }
    catch {
      case e: Exception => {
        System.out.println("Unable to create SSL Context. Reason : " + e.getMessage)
      }
    }
    return null
  }
}

File: higgs-core/src/main/java/com/fillta/higgs/ssl/SSLContextFactory.java
package com.fillta.higgs.ssl;

import javax.net.ssl.KeyManagerFactory;
import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManagerFactory;
import java.io.FileInputStream;
import java.security.KeyStore;
import java.security.SecureRandom;

public class SSLContextFactory {


    public static SSLContext getSSLSocket(SSLConfiguration sslConfiguration) {

        boolean useTrustStore = false;
        TrustManagerFactory tmf = null;
        KeyStore trustStore;

        try {
            trustStore = KeyStore.getInstance(sslConfiguration.getTrustStoreType());
            if (sslConfiguration.getTrustStorePath() != null) {
                trustStore.load(new FileInputStream(sslConfiguration.getTrustStorePath()),
                        sslConfiguration.getTrustStorePassword() == null ? "".toCharArray() : sslConfiguration.getTrustStorePassword().toCharArray());
                tmf = TrustManagerFactory.getInstance(sslConfiguration.getTrustManagerFactoryType());
                tmf.init(trustStore);
                useTrustStore = true;
            } else {
                useTrustStore = false;
            }
        } catch (Exception e) {
            System.out.println("Unable to create TrustStore Manager. Reason: " + e.getMessage());
        }

        // Load the client's key store
        boolean useClientKeyStore = false;
        KeyManagerFactory kmf = null;
        try {

            KeyStore clientKeyStore = KeyStore.getInstance(sslConfiguration.getKeyStoreType());

            if (sslConfiguration.getKeyStorePath() != null) {
                clientKeyStore.load(new FileInputStream(sslConfiguration.getKeyStorePath()),
                        sslConfiguration.getKeyStorePassword() == null ? "".toCharArray() : sslConfiguration.getKeyStorePassword().toCharArray());

                kmf = KeyManagerFactory.getInstance(sslConfiguration.getKeyManagerFactoryType());
                kmf.init(clientKeyStore, sslConfiguration.getKeyPassword() == null ? "".toCharArray() : sslConfiguration.getKeyPassword().toCharArray());
                useClientKeyStore = true;
            } else {
                useClientKeyStore = false;
            }


        } catch (Exception e) {
            System.out.println("Unable to create KeyStore Manager. Reason: " + e.getMessage());

        }
        // Create SSL Socket Factory
        try {
            SSLContext ctx = SSLContext.getInstance(sslConfiguration.getSecurityProtocol());
            ctx.init(useClientKeyStore ? kmf.getKeyManagers() : null, useTrustStore ? tmf.getTrustManagers() : null,
                    new SecureRandom());
            return ctx;

        } catch (Exception e) {
            System.out.println("Unable to create SSL Context. Reason : " + e.getMessage());
        }

        return null;
    }
}
